### PR TITLE
Block registration based on email domain

### LINKED ISSUE #4305: [Feature Request] Block registration based on email domain
I would love to be able to block registration to my gitea instance based on the email they use to register, whether or not Gitea is set to send verification emails

### PR DESCRIPTION
**Pull request by adelowo**
Fixes https://github.com/go-gitea/gitea/issues/4305

### CODE REVIEW

**Comment by adelowo** [comment]
Can a maintainer restart the build ? ... It seems to cancel after timing out

**Comment by appleboy** [comment]
@adelowo restart the PR build.

File: modules/auth/user_form.go [hunk 1]
```diff
@@ -84,6 +87,31 @@ func (f *RegisterForm) Validate(ctx *macaron.Context, errs binding.Errors) bindi
 	return validate(errs, ctx.Data, f, ctx.Locale)
 }
 
+// IsEmailDomainWhitelisted validates that the email address
+// provided by the user matches what has been configured .
+// If the domain whitelist from the config is empty, it marks the
+// email as whitelisted
+func (f RegisterForm) IsEmailDomainWhitelisted() bool {
```
Comment (lafriks) [inline modules/auth/user_form.go:95]: I think this function should be made case-insensitive

File: custom/conf/app.ini.sample [hunk 2]
```diff
@@ -308,6 +308,9 @@ ACTIVE_CODE_LIVE_MINUTES = 180
 RESET_PASSWD_CODE_LIVE_MINUTES = 180
 ; Whether a new user needs to confirm their email when registering.
 REGISTER_EMAIL_CONFIRM = false
+; List of domain names that are allowed to be used to register on a Gitea instance
+; gitea.io example.com
+EMAIL_DOMAIN_WHITELIST=
```
Comment (lafriks) [inline custom/conf/app.ini.sample:316]: It should be coma or semicolon separated same as we have in other settings

**Review by lafriks — CHANGES_REQUESTED** [review]

File: custom/conf/app.ini.sample (hunk 2 above)
Comment (adelowo) [reply custom/conf/app.ini.sample:316]: Fixed in https://github.com/go-gitea/gitea/pull/5157/commits/4fdd1aea44222a3b74de2f69bd5c96a1e783dd47

File: modules/setting/setting.go [hunk 3]
```diff
@@ -26,7 +26,6 @@ import (
 	"code.gitea.io/gitea/modules/log"
 	_ "code.gitea.io/gitea/modules/minwinsvc" // import minwinsvc for windows services
 	"code.gitea.io/gitea/modules/user"
-
```
Comment (techknowlogick) [inline modules/setting/setting.go]: Can you add this line back in?

File: routers/user/auth.go [hunk 4]
```diff
@@ -21,7 +21,6 @@ import (
 	"code.gitea.io/gitea/modules/recaptcha"
 	"code.gitea.io/gitea/modules/setting"
 	"code.gitea.io/gitea/modules/util"
-
```
Comment (techknowlogick) [inline routers/user/auth.go]: Can you add this line back in?

File: modules/auth/user_form.go [hunk 5]
```diff
@@ -84,6 +87,31 @@ func (f *RegisterForm) Validate(ctx *macaron.Context, errs binding.Errors) bindi
 	return validate(errs, ctx.Data, f, ctx.Locale)
 }
 
+// IsEmaildomainwhitelisted validates that the email address
+// provided by the user matches what has been configured .
+// If the domain whitelist from the config is empty, it marks the
+// email as whitelisted
+func (f RegisterForm) IsEmaildomainwhitelisted() bool {
```
Comment (lafriks) [inline modules/auth/user_form.go]: Please fix function name casing `IsEmailDomainWhitelisted`

File: modules/auth/user_form.go [hunk 6]
```diff
@@ -84,6 +87,31 @@ func (f *RegisterForm) Validate(ctx *macaron.Context, errs binding.Errors) bindi
...
+	if len(setting.Service.EmailDomainWhitelist) == 0 {
+		return true
+	}
+
+	n := strings.LastIndex(f.Email, "@")
+	if n <= 0 {
+		return false
+	}
+
+	domain := f.Email[n+1:]
```
Comment (lafriks) [inline modules/auth/user_form.go]: Please add to lower case here

File: modules/auth/user_form.go [hunk 7]
```diff
@@ -84,6 +87,31 @@ func (f *RegisterForm) Validate(ctx *macaron.Context, errs binding.Errors) bindi
...
+
+	for _, v := range setting.Service.EmailDomainWhitelist {
+		if v == domain {
```
Comment (lafriks) [inline modules/auth/user_form.go]: And to lower case here for `v`

**Review by lafriks — CHANGES_REQUESTED** [review]

File: modules/auth/user_form.go (hunk 5 above)
Comment (adelowo) [reply modules/auth/user_form.go]: This was what I originally had but you requested it to be changed in  https://github.com/go-gitea/gitea/pull/5157#discussion_r228081394 ... Would fix

**Review by lafriks — APPROVED** [review]

File: modules/auth/user_form.go [hunk 8]
```diff
@@ -6,6 +6,9 @@ package auth
 
 import (
 	"mime/multipart"
+	"strings"
```
Comment (techknowlogick) [inline modules/auth/user_form.go:10]: Please add Gitea copyright to header.

Comment (adelowo) [reply modules/auth/user_form.go:10]: See https://github.com/go-gitea/gitea/pull/5157/commits/2f35804137e45ec0c94335e6de18ea1a8fa2724b

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**